File: pado-rpc/src/main/java/com/netcrest/pado/rpc/client/dna/RpcInvoker.java
package com.netcrest.pado.rpc.client.dna;

import com.netcrest.pado.data.jsonlite.JsonLite;
import com.netcrest.pado.rpc.IRpc;
import com.netcrest.pado.rpc.mqtt.ReplyKey;
import com.netcrest.pado.rpc.mqtt.RequestKey;
import com.netcrest.pado.rpc.mqtt.client.MqttJsonRpcClient;

/**
 * RpcInvoker directly invokes {@link IRpc} implementation class methods.
 * Typically, a biz wrapper class that hides the RPC details is used to invoke
 * {@link IRpc} objects, instead. Although this class can directly be used by
 * applications, it is more appropriate for remotely testing {@link IRpc}
 * classes.
 * 
 * @author dpark
 *
 */
public class RpcInvoker
{
	/**
	 * Invokes the IRpc method included in the specified parameters.
	 * 
	 * @param params
	 *            JSON object containing the following parameters:
	 *            <ul>
	 *            <li>classname: Fully-qualified IRpc implementation class
	 *            name</li>
	 *            <li>method: Name of the method to invoke</li>
	 *            <li>params: Parameters in JSON object</li>
	 *            <ul>
	 * @return Returned value of the IRpc method call, null if the class name or
	 *         method is not defined in the parameters, null if the method has
	 *         the return type of void.
	 */
	@SuppressWarnings("rawtypes")
	public Object invoke(JsonLite params)
	{
		System.out.println("RpcInvoker.invoke() entered");
		String bizImplClassName = params.getString(RequestKey.classname.name(), null);
		String method = params.getString(RequestKey.method.name(), null);
		JsonLite bizImplParams = (JsonLite) params.get(RequestKey.params.name());
		int timeout = params.getInt(RequestKey.timeout.name(), 1000);
		if (bizImplClassName == null) {
			return null;
		}
		if (method == null) {
			return null;
		}
		
		System.out.println("RpcInvoker.invoke() invoking RPC...");

		JsonLite jl = MqttJsonRpcClient.getRpcClient().execute(bizImplClassName, method, bizImplParams, timeout);
		System.out.println("RpcInvoker.invoke(): reply=" + jl);
		if (jl == null) {
			return null;
		}
		Object result = jl.get(ReplyKey.result.name());
		System.out.println("RpcInvoker.invoke(): result=" + result);
		return result;
	}
}

File: pado-rpc/src/main/python/com/netcrest/pado/rpc/client/dna/rpc_invoker.py
'''
Created on Oct 4, 2017

@author: dpark
'''
from com.netcrest.pado.rpc.rpc_shared import RpcShared
from com.netcrest.pado.rpc.util import rpc_util
from com.netcrest.pado.rpc.util.rpc_util import create_request


class RpcInvoker(RpcShared):
    '''
    RpcInvoker invokes IRpc methods.
    
    RpcInvoker directly invokes IRpc implementation class methods.
    Typically, a biz wrapper class that hides the RPC details is used to invoke
    IRpc objects, instead. Although this class can directly be used by
    applications, it is more appropriate for remotely testing IRpc
    classes.
    '''
    
    def __init__(self):
        pass
        
    def invoke(self, jparams):
        print('rpc_invoker.invoker() entered')
        if not 'classname' in jparams:
            return None
        rpc_class_name = jparams['classname']
        if not 'method' in jparams:
            return None 
        method = jparams['method']
        if not 'method' in jparams:
            rpc_params = None
        else:
            rpc_params = jparams['params']
        if not 'timeout' in jparams:
            timeout = 10000
        else:
            timeout = jparams['timeout']
        timeout_in_sec = timeout / 1000
        
        print('rpc_invoker.invoker(): timeout=' + str(timeout_in_sec))
        jrequest = create_request(rpc_class_name, method, rpc_params)
        print('rpc_invoker.invoker(): jrequest=' + str(jrequest))
        jresult = self.rpc.execute(jrequest, timeout_in_sec)
        if 'error' in jresult:
            error = jresult['error']
            return rpc_util.create_error_wrap(jerror=error)
        else:
            if 'result' in jresult:
                return jresult['result']
            else:
                return None
        
        
